plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+    // but doesn't affect correctness. It's possible to return NOT_BLOCKED when it's actually
+    // blocked in certain cases. The decision is consciously made to avoid global locking.
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (freeBuffers.isEmpty() && pendingBuffers.size() < outputPartitionCount) {
+            synchronized (this) {
+                if (blockedFuture.isDone()) {
+                    blockedFuture = new CompletableFuture<>();
+                }
+                return blockedFuture;
+            }
+        }
+        else {
+            return NOT_BLOCKED;
+        }
+    }
+
+    @Override
+    public void add(int partitionId, Slice data)
+    {
+        throwIfFailed();
+
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!finished, "already finished");
+        if (finished) {
+            return;
+        }
+
+        writers.computeIfAbsent(partitionId, this::createWriter);
+        synchronized (writers.get(partitionId)) {
+            writeToExchangeStorage(partitionId, Slices.wrappedIntArray(data.length()));
+            writeToExchangeStorage(partitionId, data);
+        }
+    }
+
+    private ExchangeStorageWriter createWriter(int partitionId)
+    {
+        URI outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return exchangeStorage.createExchangeStorageWriter(outputPath, secretKey);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeToExchangeStorage(int partitionId, Slice slice)
+    {
+        int position = 0;
+        while (position < slice.length()) {
+            SliceOutput pendingBuffer = pendingBuffers.computeIfAbsent(partitionId, ignored -> {
+                try {
+                    return freeBuffers.take();
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            });
+            int writableBytes = min(pendingBuffer.writableBytes(), slice.length() - position);
+            pendingBuffer.writeBytes(slice.getBytes(position, writableBytes));
+            position += writableBytes;
+
+            flushIfNeeded(partitionId, false);
+        }
+    }
+
+    private void flushIfNeeded(int partitionId, boolean finished)
+    {
+        SliceOutput buffer = pendingBuffers.get(partitionId);
+        if (!buffer.isWritable() || finished) {
+            if (!buffer.isWritable()) {
+                pendingBuffers.remove(partitionId);
+            }
+            writers.get(partitionId).write(buffer.slice()).whenComplete((result, throwable) -> {
+                if (throwable != null) {
+                    failure.compareAndSet(null, throwable);
```

Comment:
Maybe add `throwable` as suppressed to `failure` if `failure` is already set.

Reply:
That will require additional synchronization. From what I can tell, in such patterns we usually only catch the first failure.